### DISCUSSION TITLE
Which font variable controls the TOC font?

### DISCUSSION THREAD

**Discussion by bendtherules** [koreader/koreader#8402] [Q&A] [answered]
I want to change the font-family for TOC in KoReader for kobo. The default font can't handle the conjunctives in bengali language.

In font.lua > fontmap, I have tried changing the font for following variables - cfont, tfont, smalltfont, x_smalltfont, ffont, smallffont, largeffont, rifont, pgfont, scfont. But they don't seem to control the TOC menu font. They did work for other UI parts like file browser, etc.
Please tell me where to change the font correctly for TOC.

**Comment by NiLuJe** [comment, marked as answer]
On mobile right now, but at a guess it *should* be one of the Noto Sans entry?

**Reply by bendtherules** [reply]
No hurry. I just don't want to change all of the fonts right now. Maybe if you can check the code later and update here, that'll be great.

**Reply by NiLuJe** [reply]
And the winner is... [x_smallinfofont](https://github.com/koreader/koreader/blob/d9fc34cbba48f8ac0fa3a541391f863fadc8e3a8/frontend/apps/reader/modules/readertoc.lua#L736) ;).

**Comment by poire-z** [comment]
> The default font can't handle the conjunctives in bengali language.

Seems Bengali glyphs are provided by both FreeSans and FreeSerif:

![image](https://user-images.githubusercontent.com/24273478/140163899-f9357442-3440-4b9f-a2c9-521d038c2a4c.png)

![image](https://user-images.githubusercontent.com/24273478/140163847-302be86c-c8c3-4d05-a788-9f5c85d5af1a.png)
(Sample from https://r12a.github.io/scripts/phrases, Bengali must have been added recently).

Dunno what "conjunctives" means - is one of the above better than the other ? (not sure we can swap these 2 fonts among our fallbacks without side effects on other scripts/contexts though).

Should we ship a Noto Sans Bengali ? https://en.wikipedia.org/wiki/Bengali_alphabet that's over 265 million potential users :)

**Reply by tjaderxyz** [reply]
> Should we ship a Noto Sans Bengali ? https://en.wikipedia.org/wiki/Bengali_alphabet that's over 265 million potential users :)

Isn't broad Unicode coverage the main reason the default font is Noto? In that case I don't see why any particular script wouldn't be shipped.

**Reply by poire-z** [reply]
> I don't see why any particular script wouldn't be shipped.

Because of:
1) the size of the zips we ship, which would become quite larger
2) amount of useless fonts in the `Font>` menu (would make 15 pages of `Noto Something`, like people can already witness on Android if they `[x] Enable system fonts`)
3) because FreeSans and FreeSerif provide glyphs for many other scripts (Armenian, Bengali, Cherokee, Ethiopic, Greek, Hebrew, Kanadda, Burmese, Syriac, Tifinagh, Thai...) which may be good anough for our UI (book titles, TOC) - users who want better for their book can install their own preferred font

**Reply by bendtherules** [reply]
No, both the examples you showed don't handle conjunctives. I have added a screenshot to highlight the difference below - 

**Correct** ⬇️
<img width="359" alt="Screen Shot 2021-11-04 at 6 38 24 AM" src="https://user-images.githubusercontent.com/3101150/140240630-ff78ca07-5820-4f77-8fb7-6f669d405ab7.png">

**Not correct** ⬇️
![140163847-302be86c-c8c3-4d05-a788-9f5c85d5af1a](https://user-images.githubusercontent.com/3101150/140242537-c0515537-8cbb-4f87-a385-a2b5d6cebfc8.png)


Conjunctive (or juktakkhor in bengali) literally means joined characters. So, 2 or more characters join together to form a slightly modified character (which usually joins the constituents characters vertically). For ex, in the english word *Astro* - 't', 'r' and 'o' is pronounced together, so in bengali they will be replaced with a joined character.
Now, the unsupported fonts (like the example you showed) simple render them as individual characters side-by-side with sort-of-a slanted line at the bottom indicating they should be merged, but they don't render the special joined glyph. So they are technically readable, but it would break the flow of your reading a lot since you are never used to joining them mentally and they sometimes sound different if they are joined.

In terms of support - yes, the epubs I read will have proper fonts embedded. But there are still some problems if the system font doesn't have proper support - 
1. simply finding the books in the file browser would be a chore if the filenames are in bengali too. 
2. Table of Contents is even more important, since it's technically part of the book's text - but it seems to be rendered with the system font. Is there some way for the book to suggest the font for TOC?
3. Filename also shows up in status bar, etc. (not too important)

I am not sure exactly which free fonts support it properly. But in the github phrases website you linked, all the font-families mentioned in the css support it - `Raleway, ralewaywf, adobe-clean, 'Helvetica Neue', calibri, corbel, "Segoe UI", helvetica` (but i assume they are not free to use). Mac and the Books app (all its font options) has very good support for bengali and conjunctives. On Android, Roboto also supports it.

**Reply by bendtherules** [reply]
Another broad question - is it possible that some fonts could fallback to other specialised language-specific fonts automatically depending on the unicode character? Is that what embedding "Noto Bengali" would do? Or does the user need to manually switch to it when they are reading more bengali books.

Also, is it possible make these UI fonts more configurable in KOReader without having to change the code? Something like the config variables you can already change, which need a restart to be activated?

**Comment by poire-z** [comment]
Thanks for the cultural and technical insights :)

> Is there some way for the book to suggest the font for TOC?

Nope. The embedded fonts in the EPUB are only available to the EPUB renderer (and these embedded fonts may be subsets, not having every glyphs to display other stuff which might be needed in the UI, like the TOC chevrons).

> Another broad question - is it possible that some fonts could fallback to other specialised language-specific fonts automatically depending on the unicode character? Is that what embedding "Noto Bengali" would do?

Yes.
That's what is happening with our fallback fonts - and the order we have set:
https://github.com/koreader/koreader/blob/7baeebc2d41244020cc7d8cb5249a5a64d0e401c/frontend/ui/font.lua#L105-L115
Some discussion in https://github.com/koreader/koreader/pull/3942#issuecomment-388510882 and possibly in the linked issues.
We could add and ship https://fonts.google.com/noto/specimen/Noto+Sans+Bengali (150KB x1 to x2 depending if we include bold (there is no italic) - or another variant if there is a UI one) after Noto Sans DevanagariUI - will need @NiLuJe approval, and for you to check it is correct and without bugs :)

> Or does the user need to manually switch to it when they are reading more bengali books.

We'd rather like to avoid any user manual tweak for that kind of stuff.

> Also, is it possible make these UI fonts more configurable in KOReader without having to change the code? Something like the config variables you can already change, which need a restart to be activated?

Not yet for the fallback fonts, but it could be added if needed. But the fact we have enough glyphs/scripts support in our furthest fallback fonts FreeSans and FreeSerif had avoiding the need for this until now, and you finding Bengali in FreeSans/Serif is bogus/not goo enough.
I still think that fixing this for Bengali by default is better than having the 5-10 Bengali KOReader users to have to edit their settings manually :)

**Reply by tjaderxyz** [reply]
> Nope. The embedded fonts in the EPUB are only available to the EPUB renderer (and these embedded fonts may be subsets, not having every glyphs to display other stuff which might be needed in the UI, like the TOC chevrons).

Is there some way to make the TOC use the embedded font? The TOC is in a way part of the book, so it makes sense that it would use book fonts instead of UI fonts.

**Reply by NiLuJe** [reply]
Adding the Devanagari UI set is fine by me (in fact, I think we've already discussed this in the past ;)).

**Reply by NiLuJe** [reply]
> > Nope. The embedded fonts in the EPUB are only available to the EPUB renderer (and these embedded fonts may be subsets, not having every glyphs to display other stuff which might be needed in the UI, like the TOC chevrons).
> 
> Is there some way to make the TOC use the embedded font? The TOC is in a way part of the book, so it makes sense that it would use book fonts instead of UI fonts.

I'll leave final word on that to @poire-z, but that'd be another kettle of fish entirely, by a loooong shot ;).

**Reply by poire-z** [reply]
Not an easy/obvious way. And there is no THE/single embedded font: in an EPUB, there may be multiple embedded fonts, which are associated with CSS to various HTML elements: we don't know which one is best suited for this item of the TOC - so we would need to forward them all, without any order info, and we could wrongly pick some italic or bold...
Too complicated/risky to spend more time thinking about it.  Easier to try to complete our UI font coverage, so you get correct rendering of stuff other than ToC - like Book information/description, dicts and wikipedia results (there must be a bengali Wikipedia that should then be unreadable currently).
@tjaderxyz : what language/script do you miss/have bad rendering with ? (I guess you must since you're interested on this topic :)

**Reply by tjaderxyz** [reply]
> @tjaderxyz : what language/script do you miss/have bad rendering with ? (I guess you must since you're interested on this topic :)

I only read boring Latin-script languages myself, but I do have a general interest in writing systems.

I am sure it's not something easy to implement, I'm just throwing ideas around :)

Maybe somehow putting the embedded fonts first in the fallback font list for book content would be a start. Again, I don't really know the ins and outs of the implementation, but my main point is that the TOC should ideally be treated as "book content" and not "UI content", so it should follow the same rendering rules of text inside the book.

It certainly is complicated by all the CSS interactions and stuff, but I wonder what happens if a book embeds a font and has no styling set for a given snippet. Does the embedded font get used as a fallback or is it just ignored?

**Reply by poire-z** [reply]
If there's an embedded font and it is not referenced by CSS, it won't be used anywhere.
At most, one can be set to the `<BODY>` element, so it's inherited by all the content - but a publisher does not have to do that, and may not do it: he could decide to just associate it to `<P>` paragraph nodes that usually contain the text.

**Comment by poire-z** [comment]
> Adding the Devanagari UI set is fine by me 

We already have this one :) We're talking about a Noto Sans Bengali (UI?) now :)
I'd still prefer if we had a single, possible home made, Noto Sans India/SouthEastAsia which would include the many indian scripts + the other non-CJK asian ones: they must all not have many glyphs and could fit all in a single font. This would avoid us having 10 NotoSansXYZ in the menu.

**Reply by NiLuJe** [reply]
AFAICT, the fact that Google didn't do that tells me there must be a technical reason why (as they're also using FT + HB mostly everywhere these days, the lead HB dev is a Google employee) ;).

**Reply by tjaderxyz** [reply]
Maybe the shipped fallback fonts could be hidden from the font menu to avoid polluting it, maybe even with a toggle in the developer settings to show them?

**Comment by bendtherules** [comment]
Thanks for all the clarifications.
When this gets added, i will definitely test it out on my kobo 😀

**Reply by poire-z** [reply]
Nope, we won't add it unless you have tested it :) and that it proves better than what we have with FreeSans/FreeSerif.

Can you download NotoSansBengaliUI-Regular.ttf (alone, and later NotoSansBengaliUI-Bold.ttf to check if it's really needed and faked bold is not good enough) from either:
https://github.com/googlefonts/noto-fonts/tree/main/hinted/ttf/NotoSansBengaliUI
https://github.com/googlefonts/noto-fonts/tree/main/unhinted/ttf/NotoSansBengaliUI
@NiLuJe will tell if you/we should take the hinted or unhinted one, as I don't know :)

And update manually on your device (will be reset when you update):
https://github.com/koreader/koreader/blob/7baeebc2d41244020cc7d8cb5249a5a64d0e401c/frontend/ui/font.lua#L105-L115
so it reads:
```lua
     [4] = "NotoSansDevanagariUI-Regular.ttf", 
     [5] = "NotoSansBengaliUI-Regular.ttf", 
     [6] = "nerdfonts/symbols.ttf", 
     [7] = "freefont/FreeSans.ttf", 
     [8] = "freefont/FreeSerif.ttf", 
```

**Reply by NiLuJe** [reply]
I think we got the hinted ones, but I'd have to check the commit logs.

Likely irrelevant anyway, as I doubt it's actually hinted in such a complex script ;).
(i.e., if hinting there is, it's ttfautohint).

**Reply by bendtherules** [reply]
Got it, will try these out

**Comment by poire-z** [comment]
https://notofonts.github.io/noto-docs/specimen/NotoSansBengaliUI/
So many other script specific fonts - no idea how many are not good enough with FreeSans/FreeSerif :/
May be we'd better have a menu to download these script specific fonts (UI variants) and have them inserted in this fallback list - so users can help themselves as needed, instead of us having to decide how much we allow our zip size and menu to grow :/ and privilege scripts with a large population :)

**Reply by NiLuJe** [reply]
That's... not a bad idea, but more work, of course ;).

This is essentially how it works on Kindle, FWIW. On CJK locales, the systems downloads a font package OTA ;).
